### PR TITLE
Pre-compute what's needed for AffineConstraints::distribute().

### PR DESCRIPTION
**Pull request by bangerth**
This addresses one of the things mentioned in #15375: `AffineConstraints::distribute()` requires an index set that describes what elements to import from other processes. This is the index set creation that originally tripped up @marcfehling and that led us to write #15366. This patch moves the computation of this index set into `close()`, which means that it has to be done only once for cases where one calls `distribute()` multiple times, where previously it is done in every call to `distribute()`.

There is one point where I would appreciate discussion. In the existing code, we build the index set of elements to import in the following way:
* We ask the vector on which `distribute()` is called about its locally-owned index set.
* Then we add to that all of the elements necessary to resolve constraints among the locally-owned vector elements.

The problem is that I don't have access to the locally owned set in `close()`. All I have is the locally-relevant set. So I build the index set starting from the locally-relevant set, and add what is necessary to resolve constraints within it. I end up with a bigger index set than strictly necessary, and consequently the import operation in `distribute()` will import a larger set of elements from other processes. Possible solutions to this pessimization:
* We decide that the whole optimization is simply not worth it. After #15366, the index set construction is perhaps not the bottleneck any more. (Perhaps, I have not actually measured whether that's true. I will also say that it would actually be nice to not only build the index set, but also a partitioner/communicator that we can then use in the import operation.)
* We decide that the real cost in importing elements is in the latency, and adding a few more indices is not that bad. To be clear, the additional elements are dependencies of constrained DoFs located on ghost cells but not locally owned.
* We add an argument to `close()` where callers can pass in the locally-owned index set. This allows building the same index set we do right now, at the expense of having to change our codes. We can make it backward compatible by defaulting the argument to an empty index set in which case I fall back to the algorithm with the locally-relevant index set to which I have access.

Opinions?

### CODE REVIEW

**Comment by tjhei** [comment]
We already ask the user to supply the relevant set in the call to reinit. Maybe we add an overload that also takes the locally owned set?

File: include/deal.II/lac/affine_constraints.templates.h [hunk 1]
```diff
@@ -992,6 +992,37 @@ AffineConstraints<number>::close()
...
+  // reversion of one attempt that implements this in r29662. Rather, we
+  // need to get a vector that has all the *sources* or constraints we
+  // own locally, possibly as ghost vector elements, then read from them,
+  // and finally throw away the ghosted vector. Implement this in the
+  // following.
+  needed_elements_for_distribute = local_lines;
+
+  if (needed_elements_for_distribute != complete_index_set(local_lines.size()))
+    {
+      std::vector<types::global_dof_index> additional_elements;
```
Comment (blaisb) [inline include/deal.II/lac/affine_constraints.templates.h:1015]: Not sure if this is a bottleneck, but is there any indication of a memory footprint that could be used to pre-reserve some memory? I know that we won't know which line in lines are "elements",  but maybe the size of local_lines could give us a pre-emptive indication of the memory footprint?

**Review by blaisb — COMMENTED** [review]
Just a single comment, clear implementation from my POV

File: include/deal.II/lac/affine_constraints.templates.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/lac/affine_constraints.templates.h:1015]: I don't know how to estimate this. It's the number of DoFs against which locally relevant DoFs are constrained. I don't think there is a good way of estimating this. One could duplicate the loop and in a first go-around count how many elements we get, and in the second actually insert them, calling `.reserve(...)` in between. But the calls to `is_element()` are expensive enough that I don't think that's a win :-(

File: include/deal.II/lac/affine_constraints.templates.h [hunk 2]
```diff
@@ -2264,6 +2295,8 @@ namespace internal
   } // namespace AffineConstraintsImplementation
 } // namespace internal
 
+
+
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/affine_constraints.templates.h:2299]: Just the standard 3 empty lines between function implementations.

**Comment by bangerth** [comment]
What do we do? @tjhei 's suggestion of adding another argument to the places where we currently take the `local_constraints` (=locally relevant DoF indices) is not my favorite, because we don't need this information anywhere outside of `close()`. Thus my suggestion above to only pass `locally_owned_dofs` to `close()` as a defaulted argument (and use `locally_relevant_dofs()` if not provided).

**Comment by tjhei** [comment]
I am ok with either approach, but I have two arguments for my suggestion:
1. It is more consistent to ask for both IndexSets at the same time and it also mirrors other places in the library (ghosted vectors for example).
2. With this information in place, we can potentially do other operations later: checking consistency or communicating to provide consistency.

Your approach is a bit simpler and doesn't require storing the information for later. 🤷‍♂️

**Comment by blaisb** [comment]
> I am ok with either approach, but I have two arguments for my suggestion:
> 
> 1. It is more consistent to ask for both IndexSets at the same time and it also mirrors other places in the library (ghosted vectors for example).
> 2. With this information in place, we can potentially do other operations later: checking consistency or communicating to provide consistency.
> 
> Your approach is a bit simpler and doesn't require storing the information for later. 🤷‍♂️

I would be more inclined towards this also (I like the similarity with the ghosted vectors), but both make sense to me...

**Comment by bangerth** [comment]
This seems to work finally. I'd appreciate another look.

I left a TODO in there that I'd like to leave for a follow-up patch: Even before this patch, I think we considered it a bug if you're working in parallel but didn't attach an index set for the entries to store in the constraints object. I can test this now: In the branch where we call `distribute()` on parallel vectors but with no index set given, I will have to get the MPI communicator associated with the parallel vector and test that the number of processes on this communicator is larger than one -- I think we consider this a bug, and we can assert that. I'd like to leave this for another patch because I'm pretty sure that we have tests that will break this way, and I think that deserves to be investigated separately.

**Comment by bangerth** [comment]
A separate question (also for a follow-up PR) would be whether we want to deprecate the functions that take only one index set?

**Comment by tjhei** [comment]
> I think we consider this a bug, and we can assert that.

I agree that this is a bug and we should assert. While one could argue that one can run MPI parallel computations without using any distributed data, it doesn't scale beyond a dozen or so cores and this is an easy thing to fix.

**Comment by tjhei** [comment]
> A separate question (also for a follow-up PR) would be whether we want to deprecate the functions that take only one index set?

I agree.

File: include/deal.II/lac/affine_constraints.h [hunk 3]
```diff
@@ -567,18 +567,56 @@ class AffineConstraints : public Subscriptor
...
+   * a bug.
+   *
+   * @note This constructor is equivalent to calling the following one with
+   *   both of its arguments equal to the index set provided here. This
+   *   is not wrong, but inefficient. Use the following constructor instead.
+   */
+  explicit AffineConstraints(const IndexSet &locally_stored_constraints);
+
+  /**
+   * Constructor. The supplied IndexSet defines for which indices
```
Comment (tjhei) [inline include/deal.II/lac/affine_constraints.h:594]: You now have 2 arguments. Can you update the description?

File: include/deal.II/lac/affine_constraints.h [hunk 4]
```diff
@@ -567,18 +567,56 @@ class AffineConstraints : public Subscriptor
...
+   *
+   * The @p locally_stored_constraints IndexSet allows the AffineConstraints
+   * container to save memory by just not caring about degrees of freedom
+   * that are not of importance to the current processor. In contrast,
+   * in parallel computations, if you do not provide such an index set
+   * (here, or using the reinit() function that takes such an argument), the
+   * current object will allocate memory proportional to the *total*
+   * number of degrees of freedom (accumulated over all processes),
+   * which is clearly wasteful and not efficient -- and should be considered
+   * a bug.
```
Comment (tjhei) [inline include/deal.II/lac/affine_constraints.h:616]: Can you add the advice to use the locally relevant set here?

File: include/deal.II/lac/affine_constraints.templates.h [hunk 5]
```diff
@@ -993,6 +993,38 @@ AffineConstraints<number>::close()
          ExcMessage("The constraints represented by this object have a cycle. "
                     "This is not allowed."));
 
+  // Now also compute which indices we need in distribute().
+  //
+  // This processor owns only part of the vectors we will work on. one may think
```
Comment (tjhei) [inline include/deal.II/lac/affine_constraints.templates.h:998]:
```suggestion
  // This processor owns only part of the vectors we will work on. One may think
```

File: include/deal.II/lac/affine_constraints.templates.h [hunk 6]
```diff
@@ -993,6 +993,38 @@ AffineConstraints<number>::close()
...
+  // that every processor should be able to simply communicate those elements it
+  // owns and for which it knows that they act as sources to constrained DoFs to
+  // the owner of these DoFs. This would lead to a scheme where all we need to
+  // do is to add some local elements to (possibly non-local) ones and then call
+  // compress().
+  //
+  // Alas, this scheme does not work as evidenced by the disaster of bug
+  // #51, see http://code.google.com/p/dealii/issues/detail?id=51 and the
```
Comment (tjhei) [inline include/deal.II/lac/affine_constraints.templates.h:1006]: this link doesn't exist anymore.

**Review by kronbichler — APPROVED** [review]
Apart from the comments by @tjhei, I agree with this PR and the additional constructors. However, I believe that we should try to improve the performance of the actual code even more. More precisely, the data structure that we set up is a `Utilities::MPI::NoncontiguousPartitioner` https://www.dealii.org/current/doxygen/deal.II/classUtilities_1_1MPI_1_1NoncontiguousPartitioner.html and we should take advantage of it somehow in `distribute`: We do not actually need to set up a complete vector with the entries, but we could simply use the partitioner (to be re-used between multiple calls) and some view into local arrays. Once we are at that stage, we could do the whole work in local index space. There are a few steps to do along this way, and this is one step in the right direction.

File: include/deal.II/lac/affine_constraints.templates.h (hunk 6 above)
Comment (bangerth) [reply include/deal.II/lac/affine_constraints.templates.h:1006]: Yes, I don't know what to do with this. I moved this code from another place. I replaced the link by text saying that the history of this bug was lost in a transition.

**Comment by bangerth** [comment]
I addressed the comments.

**Comment by bangerth** [comment]
Ping?

**Comment by bangerth** [comment]
Ping?

**Review by peterrum — APPROVED** [review]
Looks good to me. Let's get this in! This is already a big step forward. At the moment, we have the index sets and recompute the partitioner at every distribute call. This could be eliminated in the next step. @kronbichler Why do you think of non-contiguous partitoner? I am happy if it is used! Is this because the non-deal vectors are more "flexible"?

**Comment by bangerth** [comment]
@peterrum For the partitioner thing, take a look (and comment on!) #15988.

**Pull request merged**